### DISCUSSION TITLE
refreshListenable with Streams

### DISCUSSION THREAD

**Discussion by jhlabs** [csells/go_router#122] [Q&A] [answered]
@csells thanks for creating this fantastic library!

I am currently moving my navigation logic to `go_router` and am getting stuck on automatic redirection when my authentication changes. I would like to use the `refreshListenable` property to notify `go_router` when my authentication state changes. However, I am using the popular `bloc` library that does not expose a `Listenable` object, but instead a stream that updates as its state changes.

Has anyone found a way to make the `refreshListenable` work with streams? Would it make sense to adopt the API of `go_router` to also support streams?

**Comment by csells** [comment]
The model of a `Listenable` is to be notified of a change, to trigger an event and then to ask for the new value during the processing of that event. The model for a `Stream` is that the trigger and the value come at the same time. The way that `refreshListenable` works w/ go_router matches the `Listenable` model and not the `Stream` model, so I think the API is constructed properly. You can easily convert a `Stream` into a `Listenable` by creating an object that caches the value from the `Stream` and triggers the event for go_router.

**Reply by jhlabs** [reply]
Thanks @csells. I will try to solve it with the solution you suggested and post it here once it works with `bloc`!

**Reply by jhlabs** [reply]
I got the automatic refresh of navigation to work with this intermediate object:

```dart
class AuthStateNotifier extends ChangeNotifier {
  late final StreamSubscription<AuthState> _blocStream;
  AuthStateProvider(AuthBloc bloc) {
    _blocStream = bloc.stream.listen((event) {
      notifyListeners();
    });
  }

  @override
  void dispose() {
    super.dispose();

    _blocStream.cancel();
  }
}
```
Will mark this discussion as answered and add an example PR with bloc.

**Comment by csells** [comment]
Cool! If you wanted to contribute a bloc example, that would be cool something to add to the package.

**Reply by jhlabs** [reply]
Absolutely, once I got the solution running I can prepare an example with bloc and submit a PR.

**Reply by Pingear** [reply]
Hey)
Perhaps I'm missing something, but wrapping everything into another provider just because refreshListenable is listening is not a pretty and cumbersome idea. Especially considering the bloc is being used. Perhaps you can suggest a more elegant solution? That would be great)

@jhlabs can you show a more complete example please?)

**Reply by jhlabs** [reply]
@Pingear I am not "wrapping everything into another provider". It's a `ChangeNotifier`. The  `go_router` API exposes a `Listenable` interface, which is a reasonable choice to add a listener for route changes. My example above adapts the bloc library to this interface in very few lines of code and I personally do think it is pretty and straightforward 😉.

I think you have two choices from here on out. You can use a similar setup like I do to use this excellent routing library or you can use a different one that works in a more direct way with streams and thereby `bloc`.

**Reply by Pingear** [reply]
Hi @jhlabs 
I really like the simplicity that go_router brings to the second version of routing.
I had a question specifically in the integration with the bloc.
This is how I understood your example, and as I said, we wrap everything in the ChangeNotifierProvider just so that it hears our changes and repeats the notification about it. this is clearly unnecessary, because we are already listening to all the changes from the bloc.

  Please see an example and tell me if I understood correctly how you intend to use your example above?
  
`
        void main() {
           BlocOverrides.runZoned(
            () => runApp(MyApp()),
            blocObserver: AppBlocObserver(),
          );
        }
    
    class MyApp extends StatelessWidget {
      const MyApp({Key? key}) : super(key: key);
    
      @override
      Widget build(BuildContext context) {
        return ChangeNotifierProvider<AuthStateNotifier>(
          create: (_) => AuthStateNotifier(),
          child: BlocProvider(create: (_) => AppBloc(), child: AppView()),
        );
      }
    }
    
    class AppView extends StatelessWidget {
      AppView({Key? key}) : super(key: key);
    
      @override
      Widget build(BuildContext context) {
        var appBloc = BlocProvider.of<AppBloc>(context, listen: false);
        var loginInfo = Provider.of<AuthStateNotifier>(context, listen: false);
        loginInfo.authStateProvider(appBloc);

    var _router = GoRouter(
      refreshListenable: loginInfo,
      redirect: (state) {
        final loggedIn = appBloc.state.status == AppStatus.authenticated;
        final logging = state.location == '/login';
        if (!loggedIn && !logging) return '/login';
        if (loggedIn && logging) return '/';
        return null;
      },
      urlPathStrategy: UrlPathStrategy.path,
      routes: [
        GoRoute(
          name: 'home',
          path: '/',
          pageBuilder: (context, state) => MaterialPage(
            key: state.pageKey,
            child: HomePage(),
          ),
        ),
        GoRoute(
          name: 'login',
          path: '/login',
          pageBuilder: (context, state) => MaterialPage(
            key: state.pageKey,
            child: LoginPage(),
          ),
        ),
      ],
      errorPageBuilder: (context, state) => MaterialPage(
        key: state.pageKey,
        child: Scaffold(
          body: Center(
            child: Text(state.error.toString()),
          ),
        ),
      ),
    );


    return MaterialApp.router(
      routeInformationParser: _router.routeInformationParser,
      routerDelegate: _router.routerDelegate,
    );
  }
}

class AuthStateNotifier extends ChangeNotifier {
  late final StreamSubscription<AppState> _blocStream;

  authStateProvider(AppBloc bloc) {
    _blocStream = bloc.stream.listen((event) {
      notifyListeners();
    });
  }

  @override
  void dispose() {
    super.dispose();

    _blocStream.cancel();
  }
}

class AppBloc extends Bloc<AppEvent, AppState> {
  AppBloc() : super() {
    on<AppUserChanged>(_onUserChanged);
    on<AppLogoutRequested>(_onLogoutRequested);
  }

  void _onUserChanged(AppUserChanged event, Emitter<AppState> emit) {
    //////
  }

  void _onLogoutRequested(AppLogoutRequested event, Emitter<AppState> emit) {
    //////
  }

  @override
  Future<void> close() {
    return super.close();
  }
}`

**Comment by csells** [comment, marked as answer]
go_router provides a built-in type to support this idiom via [the `GoRouterRefreshStream` type described in the docs](https://gorouter.dev/redirection#refreshing-with-a-stream).

**Reply by Pingear** [reply]
Thank You @csells , that's cool !!

**Comment by novas1r1** [comment]
Just because I struggled a little bit with this one I would like to add my solution real quick. Maybe it saves someone else time.

```
    class AppView extends StatelessWidget {
      @override
      Widget build(BuildContext context) {
        final authCubit = context.read<FirebaseAuthCubit>();

        late final router = GoRouter(
            redirect: (state) {
              // if the user is not logged in, they need to login
              final loggedIn = authCubit.state.authStatus == AuthStatus.authenticated;

              final loggingIn = state.subloc == '/login';

              if (!loggedIn) return loggingIn ? null : '/login';

              // if the user is logged in but still on the login page, send them to
              // the home page
              if (loggingIn) return '/dashboard';

              // no need to redirect at all
              return null;
          },
          refreshListenable: GoRouterRefreshStream(authCubit.stream),
          ...
      );
```

**Comment by csells** [comment]
Thanks for giving back!
